Rhythm/Interpolation.swift

```diff
@@ -0,0 +1,193 @@
+//
+//  Tempo.Interpolation.swift
+//  Rhythm
+//
+//  Created by James Bean on 5/23/17.
+//
+//
+
+import Foundation
+import ArithmeticTools
+
+/// Interpolation between two `Tempo` values.
+///
+/// - TODO: Consider defining multiple interpolation types (linear, exponential, etc.)
+///
+public struct Interpolation {
+
+    // MARK: - Associated Types
+    public enum EasingError: Error {
+        case badValue(String, Double)
+    }
+
+    /// Easing of `Interpolation`.
+    public enum Easing {
+
+        /// Linear interpolation.
+        case linear
+
+        /// Exponential interpolation in, with the given `exponent`.
+        case exponentialIn(exponent: Double)
+
+        /// Exponential interpolation in-out, with the given `exponent`.
+        case exponentialInOut(exponent: Double)
+
+        /// Ease in / ease out (half sine wave)
+        case sineInOut
+
+        // Custom timing function modeled with cubic Bézier curve control points in the
+        // form (x,y)
+        case custom(controlPoint1: (Double, Double), controlPoint2: (Double, Double))
+
+        /// - returns: The easing function evaluated at `x`.
+        ///
+        func evaluate(at x: Double) throws -> Double {
+            guard x >= 0 && x <= 1 else {
+                throw EasingError.badValue("Input must lie in [0, 1]", x)
+            }
+
+            switch(self) {
+
+            case .linear:
+                return x;
+
+            case .exponentialIn(let e):
+                guard e > 0 else {
+                    // - TODO: should this be caught at construct time?
+                    throw EasingError.badValue("Exponent must be positive", e)
+                }
+
+                return pow(x, e)
+
+            case .exponentialInOut(let e):
+                guard e >= 1 else {
+                    // - TODO: should this be caught at construct time?
+                    throw EasingError.badValue("Exponent must be at least 1", e)
+                }
+
+                if x <= 0.5 {
```

Comment:
But semantically, what is it guarding against? Both cases are equally likely and valid